### DISCUSSION TITLE
How to Add custom dialog to Wix v4?

### DISCUSSION THREAD

**Discussion by kriveci** [wixtoolset/issues#8075] [Questions] [answered]
0
This is copy paste from my stackoverflow question(https://stackoverflow.com/questions/78224656/how-to-add-custom-dialog-to-wix-v4):

I would like to add a custom dialog to Wix v4. I found a lot of resources on how to achieve that in v3 but none for v4.

So far I have:

```
<UI>
        <ui:WixUI Id="WixUI_InstallDir" />

        <!-- Define your custom dialog -->
        <Dialog Id="MyCustomDialog" Width="370" Height="270" Title="Advanced Options">
            <Control Id="Checkbox1" Type="CheckBox" X="20" Y="100" Width="290" Height="18"
                     Property="CHECKBOX1" Text="Option 1 Description" CheckBoxValue="1" />
            <Control Id="Checkbox2" Type="CheckBox" X="20" Y="120" Width="290" Height="18"
                     Property="CHECKBOX2" Text="Option 2 Description" CheckBoxValue="1" />

            <Control Id="OkButton" Type="PushButton" X="236" Y="243" Width="56" Height="17"
                     Default="yes" Text="OK" />
            <Control Id="CancelButton" Type="PushButton" X="304" Y="243" Width="56" Height="17"
                     Cancel="yes" Text="Cancel" />
        </Dialog>

        <!-- Navigation: Insert your custom dialog into the sequence -->
        <Publish Dialog="InstallDirDlg" Control="Next" Event="NewDialog" Value="MyCustomDialog" Condition="1"/>
        <Publish Dialog="MyCustomDialog" Control="Back" Event="NewDialog" Value="InstallDirDlg" Condition="1"/>
        <Publish Dialog="MyCustomDialog" Control="Next" Event="NewDialog" Value="VerifyReadyDlg" Condition="1"/>
    </UI>
```
But MyCustomDialog is not showing. Any ideas?

**Comment by robmen** [comment]
MSI UI customization hasn't changed in v4 from v3. So, the same information should be applicable.

**Comment by chrpai** [comment, marked as answer]
My FOSS project IsWiX has WiX templates for v3 and v4 that include just this. You simply uncomment one line of XML and a custom dialog stubbed out with a header, body, footer and logic for the RememberProperty pattern is automatically inserted.

https://github.com/iswix-llc/iswix-tutorials/blob/master/WiX-v4-HeatWaveBuildTools/desktop-application/Installer/DesktopApplication/UI-CustomDialog.wxs

You simply reference the UI element.  

https://github.com/iswix-llc/iswix-tutorials/blob/master/WiX-v4-HeatWaveBuildTools/desktop-application/Installer/DesktopApplication/UI.wxs

Note the publish dialog names are dependency on which UI experience you pull in.  In this example it's made for WiXUI_FeatureTree.  This is the part you need to understand.   You need to know the name of the dialog before and after and craft the ControlEvent entries so that you override the existing behavior.

https://learn.microsoft.com/en-us/windows/win32/msi/controlevent-table

**Reply by kriveci** [reply]
Thank you. With help from your code, I made it work.
